### ISSUE TITLE
Changing default value to nil for TypeInt in Optional parameter [ Custom Providers Development ]

### ISSUE THREAD

**Issue by jeyanthimuthuram** [hashicorp/terraform#16796]
Hi,

### Questions 
* Is there any way to set default value of `TypeInt` to `nil` in optional parameters ? 

* Is there any way to find the value is provided by user or it is auto-set ?

### Terraform Version
Terraform v0.10.8

### Use Case
**I am building custom terraform provider ,there I need to call api's through the configuration** 

* I have some default values , that is optional I and even if I am passing nothing in tf configuration it is making them zero.
*  And I can't able to filter while making json request ( because that field may or may not hold zero , and I don't know any way to differentiate between user input and auto-set value)

Please suggest a way I can make it work ?

**Comment by apparentlymart** [comment]
Hi @jeyanthimuthuram,

At the moment the provider framework does not have any concept of "null", because Terraform itself doesn't either. However, work is in progress to address this via a new type system that's being introduced as part of the new version of the configuration language. It will, unfortunately, take a few steps before we reach the point where we can expose this capability to the provider framework (it has to be supported throughout Terraform Core first) but we _are_ planning to do this, for the reason you stated.

In the mean time, existing resources usually either use the zero value as an alias for "unset" or, when that isn't possible, make the attribute required and have the user set the default explicitly, or use some sort of "impossible value" as the default value, such as using `-1` for a number that can only be positive in the remote API.

**Comment by ahl** [comment]
Is the new type system still in the works? If not, it seems like there's a simple, if inelegant, fix for this problem.

**Comment by mildwonkey** [comment]
Yes, we are hard at work on the new type system. It's going to be the terraform 0.12 release, which you can read about in [the terraform blog](https://www.hashicorp.com/blog/terraform-0-1-2-preview).

**Comment by apparentlymart** [comment]
An additional note on that, though: The Terraform 0.12 release will introduce the concept of "null" at the _core_ level, but in order to preserve compatibility with older releases the provider SDK will not immediately get such support: for some time period, providers will need to work with both Terraform 0.12 and prior versions to allow for gradual upgrades.

We do still intend to add support for null attributes at the SDK level, but this will be a separate change to the SDK itself (driven by our other team that maintains the SDK) that will come at a later point where new provider releases will no longer be compatible with Terraform versions earlier than 0.12. The exact schedule for that is not yet determined, since we want to wait and see how the upgrade process for 0.12 itself goes before imposing an effective deadline on it.

**Comment by ahl** [comment]
0.12 looks pretty great: love the for loops and conditionals. Good to know that the new type system won't be immediately available to providers.
